cpp/src/arrow/type_traits.h

```diff
@@ -299,6 +299,20 @@ struct TypeTraits<ListType> {
   using ArrayType = ListArray;
   using BuilderType = ListBuilder;
   using ScalarType = ListScalar;
+  using OffsetType = Int32Type;
```

Comment:
Should be linked with `ListType::offset_type` (and CTypeTraits)? One should derive the other.

Reply:
Do you have something specific in mind?

Reply:
This is a  small nit since there's almost 99.99% chance we'll never change this.

```
using OffsetType = CTypeTraits<ListType::offset_type>
using OffsetArrayType = TypeTraits<OffsetType>::ArrayType
...
```

Reply:
Hmm, but I'm still not clear what change you're suggesting here :-)

Reply:
You're defining that the offset_type is int32_t in two places: here and type.h, my suggestion make the one in type.h authoritative.

Reply:
Or perhaps it's the snippet you posted above?

Reply:
Ah, ok, thanks.

Reply:
Well, I think it's ok like that (and a bit more readable too: you don't have to walk a chain of lookups to find out what the actual type is).

Reply:
your call, I also agree it's more readable, just slightly more prone to be forgotten in case of refactor (which has a low change of happening).